base/indices.jl

```diff
@@ -165,6 +165,9 @@ function promote_shape(a::Dims, b::Dims)
     return a
 end
 
+function promote_shape(a::Array{A}, b::Array{B}) where {A,B}
```

Comment:
As discussed in the original issue, I don't think that special-casing on `Array` is a good idea here; the error message should be generic.

Reply:
Yes, surely you want to specialise `promote_shape(axes::Base.OneTo...)` to pint differently, not `::Array`. (As a trivial example,`transpose(rand(3,3))` has the same axes but isn't an `::Array`. 

Also, if messing with this, I think the error ought to say "size" and "axes", not "dims". I presume that's an ancient message from before the terminology was standardised.

Reply:
Actually we have to stay consistent and that means:
Like in the case:
`randn(2,1)*randn(2,1)`
Throw a beautiful error:
`ERROR: DimensionMismatch("A has dimensions (2,1) but B has dimensions (2,1)")`

We should get in case of this:
`randn(3,4)+randn(3,2)`
`ERROR: LoadError: DimensionMismatch("dimensions must match: a has dims (Base.OneTo(3), Base.OneTo(2)), b has dims (Base.OneTo(3), Base.OneTo(4)), mismatch at 2")`
Instead of this:
`to this: ERROR: LoadError: DimensionMismatch("dimensions must match: a has dims (3, 2), b has dims (3, 4)"), mismatch at 2`

I face with dimensions mismatch a lot due to my work as a data scientist and it is really frustrating to find the mismatches at 4-5 level of dimensions. I can say this improvement is really simplifying it to the "python" level.

Reply:
I think the suggestion there means we should consider altering the `DimensionMismatch` to contain the dims object itself (instead of a string), and then add a custom errorshow function for it that works to consolidate this logic. Alternatively (perhaps simpler?), we could extract the code that prints the mismatch description into a separate function, so that we make the consistent error message that you mentioned from more places

Reply:
Can you clarify what exactly you want. Like can you show me the output you expect please to be more clear in the example showed above?